daemon: move base program compile to loader pkg

ianvernon opened the pull request:
This package is responsible for the management and compilation of datapath
programs for Endpoints, so logically collocating the compilation of base
programs within said package makes sense from an organizational perspective in
the code. Along with this, move the analysis of requirements for running the
daemon into `pkg/datapath/linux`.

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9488)
<!-- Reviewable:end -->

Commit:
daemon: move base program compile to loader pkg

This package is responsible for the management and compilation of datapath
programs for Endpoints, so logically collocating the compilation of base
programs within said package makes sense from an organizational perspective in
the code. Along with this, move the analysis of requirements for running the
daemon into `pkg/datapath/linux`.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -0,0 +1,303 @@
+		if err != nil {
+			return err
+		}
+	}
```
The above bits feel like they should be able to live entirely in the iptables rule manager.

joestringer replied:
(as it turns out, the proxy bits are also about managing iptables but I guess you'd have to come up with a wrapper type somewhere if we wanted to combine these two in the inputs to this function)

ianvernon replied:
Agreed - I think the proxy bits should be taken care of entirely by some entity which manages iptables (which is another refactor).

joestringer commented on pkg/datapath/loader/base.go:
```diff
@@ -0,0 +1,303 @@
+}
+
+// CompileBasePrograms compiles the base programs managed by this loader.
+func (l *Loader) CompileBasePrograms(o datapath.BaseProgramOwner, deviceMTU int, iptMgr datapath.RulesManager, p datapath.Proxy, r datapath.RouteReserver) error {
```
Given that there's a bit more than just compiling base programs (I see a rules manager, a proxy and a routeReserver), this seems almost like a:

```suggestion
// Reinitialize (re-)configures the base datapath configuration including global BPF programs, netfilter rule configuration and reserving routes in IPAM for locally detected prefixes. It may be run upon initial Cilium startup, after restore from a previous Cilium run, or during regular Cilium operation.
func (l *Loader) Renitialize(o datapath.BaseProgramOwner, deviceMTU int, iptMgr datapath.RulesManager, p datapath.Proxy, r datapath.RouteReserver) error {
```

(Although I wonder if the IPAM bits are only here because it was a convenient place for it when this was all in daemon; I'm not sure that it's logically actually part of the datapath).

ianvernon replied:
Had to rebase, so used your suggestion.

joestringer reviewed (approved):
Looks good!

I have some minor suggestions but as far as refactors go, this is nice and straightforward.

Commit:
linux: add doc for `CheckMinRequirements`

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
datapath: rename `CompileBasePrograms`

Rename to `Reinitialize` as implementations do more than just base program
compilation.

Signed-off by: Ian Vernon <ian@cilium.io>
